File: src/claw_mem/extraction/skill_store.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
claw-mem Skill Store (v4.11.0)

In-memory storage for extracted skills. Follows the lazy-load pattern:
no persistence by default. Can be extended with file persistence in
future versions.
"""

from __future__ import annotations

import uuid
from typing import Dict, List, Optional

from .skill_extractor import Skill


class SkillStore:
    """In-memory store for Skill objects.

    Stores skills keyed by a generated ID and indexed by name for
    fast lookup and deduplication (automatic merge on same-name skills).

    Usage::

        store = SkillStore()
        sid = store.store(skill)
        found = store.get(sid)
        results = store.search("management")
    """

    def __init__(self):
        self._skills: Dict[str, Skill] = {}  # id -> Skill
        self._name_index: Dict[str, str] = {}  # name (lower) -> id

    # ── CRUD ───────────────────────────────────────────────────────

    def store(self, skill: Skill) -> str:
        """Store a skill. If a skill with the same name exists, merge them.

        Args:
            skill: Skill object to store.

        Returns:
            skill_id: Unique skill identifier string.
        """
        name_key = skill.name.lower()

        if name_key in self._name_index:
            # Merge with existing
            existing_id = self._name_index[name_key]
            existing = self._skills[existing_id]
            merged = self._merge(existing, skill)
            self._skills[existing_id] = merged
            return existing_id

        # New skill
        skill_id = str(uuid.uuid4())[:8]
        self._skills[skill_id] = skill
        self._name_index[name_key] = skill_id
        return skill_id

    def get(self, skill_id: str) -> Optional[Skill]:
        """Retrieve a skill by ID.

        Args:
            skill_id: Skill identifier.

        Returns:
            Skill object or None if not found.
        """
        return self._skills.get(skill_id)

    def delete(self, skill_id: str) -> bool:
        """Delete a skill by ID.

        Args:
            skill_id: Skill identifier.

        Returns:
            True if deleted, False if not found.
        """
        skill = self._skills.pop(skill_id, None)
        if skill is None:
            return False

        name_key = skill.name.lower()
        if self._name_index.get(name_key) == skill_id:
            del self._name_index[name_key]

        return True

    def list_all(self) -> List[Skill]:
        """Return all stored skills.

        Returns:
            List of Skill objects.
        """
        return list(self._skills.values())

    def search(self, keyword: str) -> List[Skill]:
        """Search skills by keyword matching name and applicability.

        Args:
            keyword: Search term (case-insensitive substring match).

        Returns:
            List of matching Skill objects.
        """
        kw = keyword.lower()
        results: List[Skill] = []

        for skill in self._skills.values():
            if kw in skill.name.lower() or kw in skill.applicability.lower():
                results.append(skill)

        return results

    def count(self) -> int:
        """Return the number of stored skills."""
        return len(self._skills)

    def clear(self) -> None:
        """Remove all stored skills."""
        self._skills.clear()
        self._name_index.clear()

    # ── merge logic ────────────────────────────────────────────────

    @staticmethod
    def _merge(existing: Skill, incoming: Skill) -> Skill:
        """Merge two skills with the same name.

        Strategy:
          - steps: union (preserving order, removing duplicates)
          - applicability: use the longer/more specific one
          - confidence: weighted average by source_triplets
          - source: "merged"
          - other fields: from the more recent skill
        """
        # Merge steps (order-preserving union)
        seen_steps: set = set()
        merged_steps: List[str] = []
        for step in existing.steps + incoming.steps:
            if step not in seen_steps:
                seen_steps.add(step)
                merged_steps.append(step)

        # Applicability: use the longer one
        if len(incoming.applicability) > len(existing.applicability):
            applicability = incoming.applicability
        else:
            applicability = existing.applicability

        # Confidence: weighted average
        total = existing.source_triplets + incoming.source_triplets
        if total > 0:
            confidence = round(
                (existing.confidence * existing.source_triplets
                 + incoming.confidence * incoming.source_triplets)
                / total,
                2,
            )
        else:
            confidence = max(existing.confidence, incoming.confidence)

        # Source triplets: sum
        source_triplets = existing.source_triplets + incoming.source_triplets

        return Skill(
            name=existing.name,
            steps=merged_steps,
            applicability=applicability,
            confidence=confidence,
            compression_ratio=source_triplets / 1.0,
            source_triplets=source_triplets,
            created_at=min(existing.created_at, incoming.created_at),
            source="merged",
        )

File: src/extraction/skill_store.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Skill Store
 *
 * In-memory storage for extracted skills. Follows the lazy-load pattern:
 * no persistence by default.
 */

import { randomUUID } from "crypto";
import { Skill } from "./skill_extractor";

export class SkillStore {
  private _skills: Map<string, Skill> = new Map(); // id -> Skill
  private _nameIndex: Map<string, string> = new Map(); // name (lower) -> id

  // ── CRUD ─────────────────────────────────────────────────────────

  /**
   * Store a skill. If a skill with the same name exists, merge them.
   *
   * @param skill - Skill object to store
   * @returns skill_id: Unique skill identifier string
   */
  store(skill: Skill): string {
    const nameKey = skill.name.toLowerCase();

    if (this._nameIndex.has(nameKey)) {
      // Merge with existing
      const existingId = this._nameIndex.get(nameKey)!;
      const existing = this._skills.get(existingId)!;
      const merged = this._merge(existing, skill);
      this._skills.set(existingId, merged);
      return existingId;
    }

    // New skill
    const skillId = randomUUID().slice(0, 8);
    this._skills.set(skillId, skill);
    this._nameIndex.set(nameKey, skillId);
    return skillId;
  }

  /**
   * Retrieve a skill by ID.
   *
   * @param skillId - Skill identifier
   * @returns Skill object or undefined if not found
   */
  get(skillId: string): Skill | undefined {
    return this._skills.get(skillId);
  }

  /**
   * Delete a skill by ID.
   *
   * @param skillId - Skill identifier
   * @returns True if deleted, false if not found
   */
  delete(skillId: string): boolean {
    const skill = this._skills.get(skillId);
    if (skill == null) return false;

    this._skills.delete(skillId);

    const nameKey = skill.name.toLowerCase();
    if (this._nameIndex.get(nameKey) === skillId) {
      this._nameIndex.delete(nameKey);
    }
    return true;
  }

  /**
   * Return all stored skills.
   *
   * @returns Array of Skill objects
   */
  list_all(): Skill[] {
    return Array.from(this._skills.values());
  }

  /**
   * Search skills by keyword matching name and applicability.
   *
   * @param keyword - Search term (case-insensitive substring match)
   * @returns Array of matching Skill objects
   */
  search(keyword: string): Skill[] {
    const kw = keyword.toLowerCase();
    const results: Skill[] = [];
    for (const skill of this._skills.values()) {
      if (
        skill.name.toLowerCase().includes(kw) ||
        skill.applicability.toLowerCase().includes(kw)
      ) {
        results.push(skill);
      }
    }
    return results;
  }

  /** Return the number of stored skills. */
  count(): number {
    return this._skills.size;
  }

  /** Remove all stored skills. */
  clear(): void {
    this._skills.clear();
    this._nameIndex.clear();
  }

  // ── Merge logic ──────────────────────────────────────────────────

  private _merge(existing: Skill, incoming: Skill): Skill {
    // Merge steps (order-preserving union)
    const seenSteps = new Set<string>();
    const mergedSteps: string[] = [];
    for (const step of [...existing.steps, ...incoming.steps]) {
      if (!seenSteps.has(step)) {
        seenSteps.add(step);
        mergedSteps.push(step);
      }
    }

    // Applicability: use the longer one
    const applicability =
      incoming.applicability.length > existing.applicability.length
        ? incoming.applicability
        : existing.applicability;

    // Confidence: weighted average
    const total = existing.source_triplets + incoming.source_triplets;
    let confidence: number;
    if (total > 0) {
      confidence = Math.round(
        ((existing.confidence * existing.source_triplets +
          incoming.confidence * incoming.source_triplets) /
          total) *
          100,
      ) / 100;
    } else {
      confidence = Math.max(existing.confidence, incoming.confidence);
    }

    // Source triplets: sum
    const sourceTriplets = existing.source_triplets + incoming.source_triplets;

    return {
      name: existing.name,
      steps: mergedSteps,
      applicability,
      confidence,
      compression_ratio: sourceTriplets / 1.0,
      source_triplets: sourceTriplets,
      created_at: Math.min(existing.created_at, incoming.created_at),
      source: "merged",
    };
  }
}
